### PR TITLE
Sdfchoice

### PR DESCRIPTION
**Pull request by cabo**
sdfChoice proposal as discussed as a solution for #2 and #5

### CODE REVIEW

File: sdf.md [hunk 1]
```diff
@@ -829,10 +828,117 @@ numbers can be used.
...
+       "sdRef": "rgbVal"}}
+    {"type": "array", "minItems": 4, "maxItems": "4", "items": {
+       "sdRef": "cmykVal"}}
+  ]
+  ~~~
+
+  in {{-jso}} can be more descriptively notated in SDF as:
+
+  ~~~ json
+  "sdfChoice": {
```
Comment (WAvdBeek) [inline sdf.md:913]: on which level is sdfChoice being used?
for now I think it should be listed under object/parameter.
basically a replacement of JSON anyOf

Comment (cabo) [reply sdf.md:913]: At exactly the place where an "enum": [...] or an "anyOf": [...] would have been.

Comment (cabo) [reply sdf.md:913]: Details in https://raw.githack.com/ietf-wg-asdf/SDF/sdfchoice/sdf.html#name-formal-syntax-of-sdf

Please see the definition and the occurrences of `choice-type`.
JSO version of CDDL coming soooon....

**Comment by WAvdBeek** [comment]
I think we should use oneOf instead of anyOf in the syntax.
sdfChoice : selection of 1 of the choice, not 1 or more as anyOf

**Comment by cabo** [comment]
JSO oneOf has a cute name, but a bizarre semantics, which is exactly that of anyOf, but also includes a cross-check that all subschemas other than the selected one do *not* match.  Otherwise, there is no difference.  One almost never wants that bizarre semantics (although there are cases when it doesn't hurt, but still carries additional baggage).

File: sdf.md [hunk 2]
```diff
@@ -829,10 +827,117 @@ numbers can be used.
...
+
+  {{-jso}} provides a type union called `anyOf`, which provides a
+  choice between anonymous alternatives.
+
+  What could have been
+
+  ~~~ json
+  "anyOf": [
+    {"type": "array", "minItems": 3, "maxItems": "3", "items": {
+       "sdRef": "rgbVal"}}
```
Comment (akeranen) [inline sdf.md:903]: s/sdRef/sdfRef/g

File: sdf.md [hunk 3]
```diff
@@ -829,10 +827,117 @@ numbers can be used.
...
+
+~~~ json
+"enum": ["foo", "bar", "baz"]
+~~~
+
+is syntactic sugar for
+
+~~~ json
+"sdfChoice": {
+  "foo": { "const": "foo"},
```
Comment (akeranen) [inline sdf.md:935]: Would `"foo": {}` be also OK here?

**Review by akeranen — APPROVED** [review]
Small change suggestions, otherwise LGTM

File: sdf.md (hunk 3 above)
Comment (cabo) [reply sdf.md:935]: That is the little conflict between the information model character of SDF and the fact that we are expressing this in JSON instance.  By putting in the constant "foo", there is only one instance that can stand for this choice, and that is consistent with the meaning of jso's enum. The fact that a protocol binding could use the number 4711 instead doesn't really belong here; this should be done in the protocol binding.

File: sdf.md (hunk 2 above)
Comment (cabo) [reply sdf.md]: Fixed in 6ba8da2

File: sdf.md [hunk 4]
```diff
@@ -792,7 +791,6 @@ allowed for one of these types.
 | Quality          | Type                                                             | Description                                                        |
 |------------------+------------------------------------------------------------------+--------------------------------------------------------------------|
 | type             | "number" / "string" / "boolean" / "integer" / "array" / "object" | JSON data type (note 1)                                             |
-| enum             | array of allowed values                                          | enumeration constraint                                             |
```
Comment (WAvdBeek) [inline sdf.md:795]: why is this removed? I thought we agreed to keep this.

Comment (cabo) [reply sdf.md:795]: It is now in the table below this one, to keep proximity with sdfChoice.
(We could move it up again, if that is less confusing.)

Comment (cabo) [reply sdf.md:795]: (The reason I moved it down is that it is limited to strings in SDF, and therefore compatible with, but not quite the same thing as jso `enum`.)

File: sdf.md (hunk 3 above)
Comment (mjkoster) [reply sdf.md:935]: I would expect {"const": "foo") to be a popular string binding for this element, but not included in the adopted model

**Review by mjkoster — APPROVED** [review]
LGTM
